DOC - More fixes and improvements to the user guide

rcap107 opened the pull request:
This PR is fixing a number of issues in the user guide and in some other parts of the documentation, and it reworks the overall structure of the user guide to make it easier for users to parse.

The main objective was to make it easier for readers to find what they are looking for by searching the left sidebar in the user guide. The current (as time of edit) user guide has a lot of content hidden behind titles that do not really describe the content of their section. 

This PR addresses the problem by moving the content of the user guide from long files to a new `modules/` folder, which contains a subfolder for each "chapter" of the user guide. Subfolders contain the old subsections.

The result is that now the sidebar shows the title of these subsections, making it easier for users to find information about the library.

The content of the guide has remained mostly the same: I reused most of what was already there, sometimes integrating the text with examples, or moving things around for clarity.

Additional changes include:
- fixing all the broken references 
- rewriting the user guide intro 
- changing the user guide outline to go from simpler to more complex content 
- consistent use of "skrub" (rather than "Skrub" or ``skrub``) throughout the docs
- misc typo fixes
- adding an example of a data ops report

dholzmueller reviewed (commented):
Thanks, looks mostly good to me!

rcap107 commented on doc/userguide_data_cleaning.rst:
```diff
@@ -71,9 +73,29 @@ considered "uninformative":
 If you have a series containing strings with typos, the |deduplicate| function
 may be used to remove some typos by creating a mapping between the typo strings
 and the correct strings. See the documentation for caveats and more detail.
+with vim, how can I create u
```
whoops
a stray paste

rcap107 commented:
Main changes:
- I extended the section on the `SquashingScaler` and added links
- I split the user guide in subsections so that it is easier to parse for the readers
- I extended the section on the global config
- I rewrote and clarified various sections
- I fixed a few typos/formatting issues
- I changed the destination of the "Learn more about Data Ops" in the website homepage so that it goes to the example rather than the user guide. 

I still want to improve the Data Ops getting started example and user guide section.

jeromedockes commented on doc/documentation.rst:
```diff
@@ -23,16 +23,13 @@ See the examples section for full code snippets.
-   userguide_data_ops_validation
-   userguide_joining_tables
-   userguide_utils
+   :maxdepth: 3
```
still about the 1st paragraph above:
- "helps performing" -> "helps perform" (I think)
- should we sneak a very brief mention of tablereport in there?
- I guess skrub goes beyond preprocessing, as it helps build and serialize whole pipelines all the way down to the final estimator + tuning and validation. I'm not sure but maybe we can slightly reword or add a short sentence to give a more comprehensive picture, it seems this intro is only describing the tablevectorizer (and the joiners which are problematic for various reasons)

rcap107 replied:
I rewrote the intro here

jeromedockes commented on doc/documentation.rst:
```diff
@@ -23,16 +23,13 @@ See the examples section for full code snippets.
+   cleaning_a_dataframe
+   encoding_a_column
+   building_a_pipeline
+   selectors
```
the first 4 titles are great, but "selectors" and "dataops" are much less informative. could we change them to match the styles of the previous ones, describing the task rather than naming the module? I know it's hard to do in a short title

jeromedockes replied:
or maybe each section of the TOC could include a short sentence or paragraph below the title to say what this is about and who might be interested to read it

jeromedockes commented on doc/modules/tablereport/altering_appearance.rst:
```diff
@@ -0,0 +1,22 @@
+.. |TableReport| replace:: :class:`~skrub.TableReport`
```
All of the tablereport pages are very short I think they could have stayed in one page to avoid one level of nesting. there is not so much to say about the table report that it requires multiple web pages and this way we shorten the path `home > user guide TOC > tablereport TOC > tablereport` to reach the first line of actual content about the table report

jeromedockes commented on doc/modules/tablereport/exploring_dataframes_interactively.rst:
```diff
@@ -0,0 +1,48 @@
+Exploring dataframes interactively with the |TableReport|
+=========================================================
+
+The |TableReport| gives a high-level overview of the given dataframe, suitable for
+quick exploratory analysis of series and dataframes. The report shows the first
```
```suggestion
The |TableReport| gives a high-level overview of a Dataframe or Series, suitable for
quick exploratory analysis. The report shows the first
```

jeromedockes commented on doc/modules/tablereport/exploring_dataframes_interactively.rst:
```diff
@@ -0,0 +1,48 @@
+
+The |TableReport| gives a high-level overview of the given dataframe, suitable for
+quick exploratory analysis of series and dataframes. The report shows the first
+and last 5 rows of the dataframe (decided by the ``n_rows`` parameter), as well
```
an actual report would be useful here, and maybe void the need for a textual description

rcap107 replied:
is it possible to embed a report in the documentation?

jeromedockes reviewed (commented):
thanks a lot @rcap107 ! I noted a couple of details for now but I need to spend more time on it to get a complete picture :)

rcap107 commented on doc/cleaning_a_dataframe.rst:
```diff
@@ -0,0 +1,16 @@
+   modules/data_cleaning/cleaning_dataframes_and_parsing_datatypes
+   modules/data_cleaning/removing_unneeded_columns
+   modules/data_cleaning/robust_scaling
+   modules/data_cleaning/deduplicate_categorical_data
```
will do

GaelVaroquaux commented:
> We need to converge on the titles for the outline, but I think the PR is in a decent shape.

Do you think that I can review it with the goal of merging it?

rcap107 commented:
> We need to converge on the titles for the outline, but I think the PR is in a decent shape.
> Do you think that I can review it with the goal of merging it?

If you think the overall outline is good, I think so. We can always iterate and fix stuff later

rcap107 commented:
> > we really should move the dataops examples to 50+ or something, otherwise every time we change one of the previous examples we need to go and fix all the links
> 
> That's a good idea! Do you want to have a go at it?

Only after this gets merged

GaelVaroquaux commented on doc/modules/multi_column_operations/column_selection_pipeline.rst:
```diff
@@ -0,0 +1,10 @@
+Column selection inside a pipeline
+----------------------------------
```
```suggestion
----------------------------------

.. currentmodule:: skrub

```

This is important so that the symbols actually turn into links to the API documentation (very important for the user)

rcap107 commented:
> General comment: many of the `:class:...` entries are now broken. I did a suggestion to fix one (adding a "currentmodule") but we need to do it elsewhere, for instance here https://output.circle-artifacts.com/output/job/7c544d63-1587-438f-a73b-53ed1a926088/artifacts/0/doc/modules/joining_tables/interpolation_joiner.html
> 
> @rcap107 : could you review a bit the generated code to find and fix these issues. They are a quite severe regression

I'll check all the files tomorrow

jeromedockes reviewed (commented):
thanks a lot for all this @rcap107 , I'll continue reading it :)

rcap107 commented on doc/column_level_featurizing.rst:
```diff
@@ -0,0 +1,18 @@
+.. _user_guide_encoders_index:
```
yes adding numbers to the examples is something I had considered doing

rcap107 commented on doc/modules/default_wrangling/deduplicate_categorical_data.rst:
```diff
@@ -0,0 +1,49 @@
+
+.. _user_guide_deduplicate:
+
+Deduplicate categorical data with |deduplicate|
```
I can rename the config section  "configuration and utilities" and move this part there

rcap107 commented on doc/_static/dataop_report/index.html:
```diff
@@ -0,0 +1,374 @@
+<!DOCTYPE html>
```
I completely forgot that's how it was done, I fixed the links and removed those files

GaelVaroquaux commented on doc/modules/column_level_featurizing/feature_engineering_datetimes.rst:
```diff
@@ -0,0 +1,276 @@
+>>> to_dt.output_time_zone_ is None
+True
+
+Once ``ToDatetime`` was successfully fitted, ``transform`` will always try to
```
```suggestion
Once |ToDatetime| was successfully fitted, ``transform`` will always try to
```

GaelVaroquaux commented on doc/modules/column_level_featurizing/feature_engineering_datetimes.rst:
```diff
@@ -0,0 +1,276 @@
+
+Caveats when dealing with month first/day first conventions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When parsing strings in one of the formats above, ``ToDatetime`` tries to guess
```
```suggestion
When parsing strings in one of the formats above, |ToDatetime| tries to guess
```

GaelVaroquaux commented on doc/modules/multi_column_operations/multi_column_operations.rst:
```diff
@@ -0,0 +1,186 @@
+- |ApplyToCols| maps a transformer to columns in a dataframe, so that all
+  columns that satisfy a certain condition are transformed, while the others are
+  left untouched.
+- |ApplyToFrame| applies a transformer to a collection of columns _at once_.
```
```suggestion
- |ApplyToFrame| applies a transformer to a collection of columns *at once*.
```
Fixing rst

GaelVaroquaux commented on doc/modules/multi_column_operations/drop_uninformative.rst:
```diff
@@ -0,0 +1,103 @@
+0  2  4
+1  3  5
+2  2  6
+
```
```suggestion

|

```
Some vertical whitespace, for readability

GaelVaroquaux commented on doc/modules/multi_column_operations/selectors.rst:
```diff
@@ -49,16 +47,12 @@ This selector can then be passed to a :func:`~skrub.selectors.select` function:
 estimator with the :func:`skrub.DataOp.skb.apply` function:
 
 >>> import skrub
+>>> from skrub import SelectCols
```
```suggestion
```

This line is not needed here (and thus confused me) but needed below

GaelVaroquaux commented on doc/modules/multi_column_operations/selectors.rst:
```diff
@@ -68,7 +62,6 @@ Result:
 0   A4   4       -1.0      -1.0
 1   A3   3        1.0       1.0
 
```
```suggestion

Selectors can be used within the :class:`skrub.SelectCols` class, implementing `fit` and `transform`, as demoed below::

  >>> from skrub import SelectCols

```

GaelVaroquaux commented on doc/modules/multi_column_operations/type_of_selectors.rst:
```diff
@@ -0,0 +1,91 @@
+0      297.0     210.0   4
+1      420.0     297.0   3
+
+Selectors can be inverted with ``~``, or :func:`~skrub.selectors.inv`:
```
```suggestion
Selectors can be inverted with ``~``, or :func:`~skrub.selectors.inv`:

```
The lack of empty line broken the rendering

GaelVaroquaux commented on doc/modules/multi_column_operations/type_of_selectors.rst:
```diff
@@ -0,0 +1,91 @@
+   height_mm  width_mm
+0      297.0     210.0
+1      420.0     297.0
+
```
```suggestion

|

```
Vertical white space, for readability

GaelVaroquaux commented on doc/data_ops.rst:
```diff
@@ -0,0 +1,71 @@
+
+In this section we cover all about the skrub Data Ops, from starting out with a
+simple example, to more advanced concepts like parameter tuning and and pipeline
+validation.
```
The below does not render great in the site's table of contents (because the titles on this page are not exposed on the site's table of content) but I don't know how to fix this, so I'm moving on for now.

GaelVaroquaux commented on doc/modules/data_ops/basics/building_data_ops_plan.rst:
```diff
@@ -0,0 +1,94 @@
+.. currentmodule:: skrub
+
```
That's a lot of code for a documentation page. I typically prefer moving these to an example, I find it easier to maintain.

GaelVaroquaux replied:
The same comment applies to many of the pages in this section.

We could make a separate, dedicated gallery, that is inserted here, if useful

GaelVaroquaux commented on doc/modules/data_ops/validation/hyperparameter_tuning.rst:
```diff
@@ -0,0 +1,202 @@
+.. _user_guide_data_ops_feature_selection:
+
+Feature selection with skrub :class:`SelectCols` and :class:`DropCols`
+==================================================================
```
```suggestion
=======================================================================
```

GaelVaroquaux commented on doc/modules/data_ops/validation/hyperparameter_tuning.rst:
```diff
@@ -0,0 +1,202 @@
+of the search using a parallel coordinates plot.
+
+A full example of how to use hyperparameter search is available in
+:ref:`sphx_glr_auto_examples_data_ops_13_choices.py`.
```
```suggestion
:ref:`sphx_glr_auto_examples_data_ops_13_choices.py`.

|

```

Vertical white space for readability

GaelVaroquaux commented on doc/modules/data_ops/validation/nesting_choices_choosing_pipelines.rst:
```diff
@@ -0,0 +1,107 @@
+
+Also note that as seen above, choices can be nested arbitrarily. For example it
+is frequent to choose between several estimators, each of which contains choices
+in its hyperparameters.
```
```suggestion
in its hyperparameters.

|

```

Vertical whitespace, to separate out the titles
